zcash_primitives/src/transaction/sighash.rs

```diff
@@ -113,24 +57,43 @@ impl AsRef<[u8; 32]> for SignatureHash {
     }
 }
 
+/// Addtional context that is needed to compute signature hashes
+/// for transactions that include transparent inputs or outputs.
+pub trait TransparentAuthorizingContext: transparent::Authorization {
+    /// Returns the list of all transparent input amounts, provided
+    /// so that wallets can commit to the transparent input breakdown
+    /// without requiring the full data of the previous transactions
+    /// providing these inputs.
+    fn input_amounts(&self) -> Vec<Amount>;
+    /// Returns the list of all transparent input scriptPubKeys, provided
+    /// so that wallets can commit to the transparent input breakdown
+    /// without requiring the full data of the previous transactions
+    /// providing these inputs.
+    fn input_scriptpubkeys(&self) -> Vec<Script>;
+}
+
+/// Computes the signature hash for an input to a transaction, given
+/// the full data of the transaction, the input being signed, and the
+/// set of precomputed hashes produced in the construction of the
+/// transaction ID.
 pub fn signature_hash<
     'a,
+    TA: TransparentAuthorizingContext,
     SA: sapling::Authorization<Proof = GrothProofBytes>,
-    A: Authorization<SaplingAuth = SA>,
+    A: Authorization<SaplingAuth = SA, TransparentAuth = TA>,
 >(
     tx: &TransactionData<A>,
-    hash_type: u32,
     signable_input: &SignableInput<'a>,
     txid_parts: &TxDigests<Blake2bHash>,
 ) -> SignatureHash {
     SignatureHash(match tx.version {
         TxVersion::Sprout(_) | TxVersion::Overwinter | TxVersion::Sapling => {
```

Comment:
Does `v4_signature_hash` handle pre-Sapling hashes? Either it's misnamed or it is miscalculating those hashes.

Reply:
https://github.com/zcash/librustzcash/issues/483